src/Interpreters/RowRefs.cpp

```diff
@@ -96,52 +102,56 @@ const RowRef * AsofRowRefs::findAsof(Type type, ASOF::Inequality inequality, con
     return out;
 }
 
-std::optional<AsofRowRefs::Type> AsofRowRefs::getTypeSize(const IColumn * asof_column, size_t & size)
+std::optional<TypeIndex> AsofRowRefs::getTypeSize(const IColumn * asof_column, size_t & size)
 {
-    if (typeid_cast<const ColumnVector<UInt32> *>(asof_column))
-    {
-        size = sizeof(UInt32);
-        return Type::keyu32;
-    }
-    else if (typeid_cast<const ColumnVector<UInt64> *>(asof_column))
-    {
-        size = sizeof(UInt64);
-        return Type::keyu64;
-    }
-    else if (typeid_cast<const ColumnVector<Int32> *>(asof_column))
-    {
-        size = sizeof(Int32);
-        return Type::keyi32;
-    }
-    else if (typeid_cast<const ColumnVector<Int64> *>(asof_column))
-    {
-        size = sizeof(Int64);
-        return Type::keyi64;
-    }
-    else if (typeid_cast<const ColumnVector<Float32> *>(asof_column))
-    {
-        size = sizeof(Float32);
-        return Type::keyf32;
-    }
-    else if (typeid_cast<const ColumnVector<Float64> *>(asof_column))
-    {
-        size = sizeof(Float64);
-        return Type::keyf64;
-    }
-    else if (typeid_cast<const ColumnDecimal<Decimal32> *>(asof_column))
-    {
-        size = sizeof(Decimal32);
-        return Type::keyDecimal32;
-    }
-    else if (typeid_cast<const ColumnDecimal<Decimal64> *>(asof_column))
-    {
-        size = sizeof(Decimal64);
-        return Type::keyDecimal64;
-    }
-    else if (typeid_cast<const ColumnDecimal<Decimal128> *>(asof_column))
+    TypeIndex idx = columnVectorDataType(asof_column);
+    if (idx == TypeIndex::Nothing)
+        idx = columnDecimalDataType(asof_column);
```

Comment:
Maybe better to make a function like `getDataType` ?

Reply:
If you're talking about `IColumn::getDataType()`, it's a good idea. But I do not want to change all columns files in this PR.